### PR TITLE
Process and thread list improvements

### PR DESCRIPTION
**Pull request by quijote**
Introduce a new column "Up Time" in process list.
Change formatting of "Total CPU" in process and thread list to correspond to "Up Time"
Introduce column "Modules" in thread list which shows up to 10 different modules seen in stack of a thread.
If sorting by "CPU Usage", use "Total CPU" as secondary sort key.
Introduce small delay of 100 ms, if sorting by "CPU Usage" in thread list. Allows to immediately show a busy thread on top if a process is selected.

### CODE REVIEW

**Comment by CyberShadow** [comment]
Hi, just wanted to say that I did not forget about this; I am doing some work over in #40, and would like to look at this after that's done.

**Comment by quijote** [comment]
Thank you for the update. #40 looks very promising and adds to my optimism that a new and even better release of VerySleepy is on the horizon. I hope I find the time to work on a few more ideas for the UI in the next weeks.

**Comment by CyberShadow** [comment]
Hi,

Now that #40 is merged, we can look at this more closely.

First, could you rebase on top of the latest master? It looks like some of the whitespace fixes I've done caused conflicts with your changes.

I'm also seeing a few whitespace inconsistencies in the diff. The new `.gitattributes` file should flag them as errors. You can also avoid introducing whitespace errors by configuring your editor / IDE to use the our  `.editorconfig` file.

File: src/wxProfilerGUI/processlist.cpp [hunk 1]
```diff
@@ -255,6 +266,14 @@ void ProcessList::sortByTotalCpuTime()
 		std::stable_sort(processes.begin(), processes.end(), TotalCpuTimeDescPred());
 }
 
+void ProcessList::sortByUpTime()
+{
+	if (sort_dir == SORT_UP)
+		std::stable_sort(processes.begin(), processes.end(), UpTimeAscPred());
+	else
+		std::stable_sort(processes.begin(), processes.end(), UpTimeDescPred());
```
Comment (CyberShadow) [inline src/wxProfilerGUI/processlist.cpp]: Could this (and `UpTimeDescPred`) be simplified using `processes.rbegin()` and `processes.rend()` ?

File: src/wxProfilerGUI/threadlist.cpp [hunk 2]
```diff
@@ -228,11 +234,28 @@ void ThreadList::updateSorting()
...
+		__int64 upTimeS = durationMs / 1000;
+		int h = (int) (upTimeS / 3600);
+		int m = (int) ((upTimeS % 3600) / 60);
+		int s = (int) (upTimeS % 60);
+		sprintf(str, "%02d:%02d:%02d", h, m, s);
+	}
+	else {
+		strcpy(str, "-");
+	}
+	return str;
```
Comment (CyberShadow) [inline src/wxProfilerGUI/threadlist.cpp]: You are returning a pointer to a stack variable here. If this code works, it works purely by chance. Your compiler should have flagged this.

Either way, fixed-length char arrays make me nervous. Use `wxString::Format` instead, or, at least, `sprintf_s` / `snprintf`.

File: src/wxProfilerGUI/threadlist.h [hunk 3]
```diff
@@ -69,9 +70,10 @@ class ThreadList : public wxSortedListCtrl
 
 	enum {		
 		COL_LOCATION,
+		COL_MODULES,
 		COL_CPUUSAGE,
 		COL_TOTALCPU,
-		COL_ID,
+		COL_ID,
```
Comment (CyberShadow) [inline src/wxProfilerGUI/threadlist.h]: Spurious whitespace-only change.

File: src/wxProfilerGUI/threadpicker.cpp [hunk 4]
```diff
@@ -172,8 +172,9 @@ ThreadPicker::ThreadPicker()
 	buttons->Add(time_check,										0, wxALIGN_LEFT | wxALIGN_CENTER_VERTICAL,	0);
 	buttons->Add(time_ctrl,											0, wxALIGN_LEFT | wxALIGN_CENTER_VERTICAL,	0);
 	buttons->AddStretchSpacer();
-	buttons->Add(all_button,										0, wxALIGN_RIGHT,							0);
-	buttons->Add(ok_button,											0, wxALIGN_RIGHT | wxLEFT,					5);
+	buttons->Add(all_button,                                        0, 0,                                       0);
+    buttons->Add(ok_button,                                         0, wxLEFT,                                  5);
```
Comment (CyberShadow) [inline src/wxProfilerGUI/threadpicker.cpp]: Wrong indentation (spaces instead of tabs).

File: src/wxProfilerGUI/threadlist.cpp (hunk 2 above)
Comment (quijote) [inline src/wxProfilerGUI/threadlist.cpp]: I tried to closely follow the programming style used by the rest of the code. Fixed-length arrays were used there for formatting. But I agree with your assessment and would not write code like this in my own codebase :).

**Comment by quijote** [comment]
I was trying to profile a process with 20000 threads again today. Hopefully I can soon find some time to work on your remarks and prepare the changes for integration into master.

**Pull request closed without merging**